---
title: "Cross-sectional BMI, Waist and waist-hip ratio and gut microbiome"
author: "Li-Fang Yeo"
date: "28-02-2025"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```
```{r}
library(magrittr)
library(ANCOMBC)
library(mia)
library(ggplot2)
library(dplyr)
```
```{r}
devtools::load_all()
```
Command line arguments
```{r}
args <- list(
  optparse::make_option("--east", action="store_true", default=FALSE, help="Exclude Eastern Finland subpopulation [default \"%default\"]"),
  optparse::make_option("--west", action="store_true", default=FALSE, help="Exclude Eastern Finland subpopulation [default \"%default\"]"),
  optparse::make_option("--low", action="store_true", default=FALSE, help="Exclude low GFR [default \"%default\"]"),
  optparse::make_option("--high", action="store_true", default=FALSE, help="Exclude high GFR [default \"%default\"]"),
  optparse::make_option("--detection", type = "numeric", default = 0.1/100, help = "Detection limit [default %default]"),
optparse::make_option("--prevalence", type = "numeric", default = 5/100, help = "Prevalence limit [default %default]")) %>% 
  optparse::OptionParser(option_list = .) %>%
  optparse::parse_args()
```
```{r}
args %>% tibble::enframe(name = "Option", value = "Argument") %>% DT::datatable()
```
Data definition
```{r}
vars <- list(BL_AGE = "Age",
             MEN = "Men",
             BMI = "BMI",
             VYOTARO = "Waist circ",
             WHR = "waist-hip ratio",
             PREVAL_DIAB = "Diabetes",
             PREVAL_CVD = "Cardiovascular disease",
             CURR_SMOKE = "Smoking",
             ALKI2_FR02 = "Alcohol consumption per week (g)",
             HFC = "Healthy food choices",
             Q57 = "Exercise",
             observed = "Observed richness",
             shannon = "Shannon diversity")
```
Read data in and filter for covariates
```{r}
tse <- readRDS("../data/tse_mgs-20241118_104759.rds") %>%
  mia::transformAssay(assay.type = "counts", method = "relabundance") %>% 
  mia::addAlpha(assay.type = "counts", index = c("shannon", "observed"), name = c("shannon", "observed"), niter = 10) %>%
  tse_add_food_score(HFC) %>%
  #combine values into one
  tse_mutate(Q57 = dplyr::if_else(Q57 %in% c(3, 4), 3, Q57)) %>%
  tse_mutate(dplyr::across(c(MEN,
                             CURR_SMOKE,
                             EAST,
                             Q57,
                             dplyr::contains("INCIDENT"),
                             dplyr::contains("PREVAL")), as.factor)) %>% 
  #pregnant
  tse_filter(GRAVID %in% c(1, NA)) %>%
  #antibiotic use in the past 1 month
  tse_filter(BL_USE_RX_J01_1mo %in% c(0, NA)) %>%
  #missing covariate
  tse_filter(dplyr::if_all(dplyr::one_of(names(vars)), not_na)) %>%
  #low reads
  tse_filter(total_reads > 50000) %>%
  tse_select(names(vars))
```

# Characteristic
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>%
  mytableone(vars)
```
# Model function definitions
BMI model + covariates
```{r}
lm_bmi_partial <-  purrr::partial(lm,
                                  formula = scale(BMI) ~
                                    scale(term) + BL_AGE + MEN +
                                    PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57)
```
```{r}
lm_bmi_model_for_taxon <- function(df, term, rank = NULL) {
  message(term)
  { if (is.null(rank)) df else df[[rank]] } %>%
    dplyr::rename(term := {{term}}) %>% 
    lm_bmi_partial(data = .)
}
```
Waist circ. model + covariates
```{r}
lm_waist_partial <-  purrr::partial(lm,
                                  formula = scale(VYOTARO) ~
                                    scale(term) + BL_AGE + MEN +
                                    PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57)
```
```{r}
lm_waist_model_for_taxon <- function(df, term, rank = NULL) {
  message(term)
  { if (is.null(rank)) df else df[[rank]] } %>%
    dplyr::rename(term := {{term}}) %>% 
    lm_waist_partial(data = .)
}
```
WHR model + covariates
```{r}
lm_whr_partial <-  purrr::partial(lm,
                                  formula = scale(WHR) ~
                                    scale(term) + BL_AGE + MEN +
                                    PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57)
```
```{r}
lm_whr_model_for_taxon <- function(df, term, rank = NULL) {
  message(term)
  { if (is.null(rank)) df else df[[rank]] } %>%
    dplyr::rename(term := {{term}}) %>% 
    lm_whr_partial(data = .)
}
```
# Alpha diversity using microbial counts on Shannon index
## Univariable adjusted for age and sex
BMI
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(BMI) ~ scale(shannon) + BL_AGE + MEN, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to BMI")
```
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(BMI) ~ scale(observed) + BL_AGE + MEN, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to BMI")
```

```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(BMI) ~ scale(shannon) + BL_AGE + MEN + PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to BMI")
```
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(BMI) ~ scale(observed) + BL_AGE + MEN + PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to BMI")
```

Waist circumference
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(VYOTARO) ~ scale(shannon) + BL_AGE + MEN, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to Vyot")
```
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(VYOTARO) ~ scale(observed) + BL_AGE + MEN, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to Vyot")
```

```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(VYOTARO) ~ scale(shannon) + BL_AGE + MEN + PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to Vyot")
```
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(VYOTARO) ~ scale(observed) + BL_AGE + MEN + PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to Vyot")
```

Hip to waist ratio
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(WHR) ~ scale(shannon) + BL_AGE + MEN, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to WHR")
```
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(WHR) ~ scale(observed) + BL_AGE + MEN, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to WHR")
```

```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(WHR) ~ scale(shannon) + BL_AGE + MEN + PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to WHR")
```
```{r}
tse %>%
  tse_meta(rownames = FALSE) %>% 
  lm(scale(WHR) ~ scale(observed) + BL_AGE + MEN + PREVAL_DIAB + PREVAL_CVD + CURR_SMOKE +
                                    ALKI2_FR02 + HFC + Q57, data = .) %>% 
  broom::tidy(conf.int = TRUE) %>%
  round_numeric_columns() %>% 
  DT::datatable(caption = "Alpha diversity to WHR")
```

#Common taxa
Agglomerate prevalent features
```{r}
#reads were filtered to detection threshold of 0.1% relative abundance, in 5% of samples
#This is to filter out rare taxa, to ensure we are anlaysing microbes that are likely to be real and not sequencing artefacts
subset_features_partial <-  purrr::partial(mia::agglomerateByPrevalence,
                                           x = tse,
                                           detection = args$detection,
                                           prevalence = args$prevalence,
                                           as_relative = TRUE) %>%
  purrr::possibly()
```
Subset to only Species
```{r}
taxa_subsets <- c("Species") %>%
  rlang::set_names() %>%
  purrr::map(subset_features_partial)
```
Melt tse to data frame
```{r}
melt_tse <- function(x, method = "clr") {
  mia::transformAssay(x, method = method, pseudocount = 1) %>% 
    mia::meltSE(add_row_data = TRUE, assay_name = method) %>%
    dplyr::mutate(FeatureID = glue::glue("GUT_{FeatureID}")) %>%
    dplyr::mutate(FeatureID = stringr::str_replace_all(FeatureID, c(" " = "_", "-" = "_"))) %>% 
    dplyr::select(SampleID, FeatureID, clr) %>%
    tidyr::spread(FeatureID, clr) %>% 
    dplyr::full_join(tse_meta(x), by = dplyr::join_by(SampleID == rownames))
}
```
```{r}
dfs <- taxa_subsets %>%
  purrr::map(melt_tse, .progress = TRUE)
```
# Taxa level analyses
generate a list of taxa, add the word GUT in front of the taxa
```{r}
taxa_rank_list <- function(x, y) {
  list_names <- colnames(x) %>%
    stringr::str_subset("GUT_") %>%
    rlang::set_names()
}
```
```{r}
taxa_rank_list <- dfs %>%
  purrr::imap(taxa_rank_list, .progress = TRUE)
```
## BMI
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_bmi_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_bmi_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_bmi_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for BMI")
```
## WaistCirc
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_waist_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_waist_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_waist_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for Waist Circ")
  
```
## Waist to hip ratio
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_whr_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_whr_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_whr_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for Waist Hip ratio")
```
Top 15 significant taxa
```{r}
whr<- df_lm_whr_results %>%
  dplyr::filter(qval_fdr < 0.05) %>%
  dplyr::arrange(qval_fdr) %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Species:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("WHR", 15))

bmi <- df_lm_bmi_results %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Species:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("BMI", 15))

waist <- df_lm_waist_results %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Species:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("waist", 15))

hi <- rbind(whr, bmi, waist)

#define colors
my_colors <- c("waist" = "darkgreen", "BMI" = "orange", 
               "WHR" = "thistle")

q <- ggplot(data=hi, aes(x=taxa, y=estimate, fill = Origin)) +
  geom_bar(stat="identity", position = 'dodge') +
  scale_fill_manual(values=my_colors) +
  coord_flip() +
  theme_classic() +
  theme(axis.text.y = element_text(face = 'italic'))
q
```

#Common taxa
Agglomerate prevalent features
```{r}
#reads were filtered to detection threshold of 0.1% relative abundance, in 5% of samples
#This is to filter out rare taxa, to ensure we are anlaysing microbes that are likely to be real and not sequencing artefacts
subset_features_partial <-  purrr::partial(mia::agglomerateByPrevalence,
                                           x = tse,
                                           detection = args$detection,
                                           prevalence = args$prevalence,
                                           as_relative = TRUE) %>%
  purrr::possibly()
```
Subset to only Genus
```{r}
taxa_subsets <- c("Genus") %>%
  rlang::set_names() %>%
  purrr::map(subset_features_partial)
```
Melt tse to data frame
```{r}
melt_tse <- function(x, method = "clr") {
  mia::transformAssay(x, method = method, pseudocount = 1) %>% 
    mia::meltSE(add_row_data = TRUE, assay_name = method) %>%
    dplyr::mutate(FeatureID = glue::glue("GUT_{FeatureID}")) %>%
    dplyr::mutate(FeatureID = stringr::str_replace_all(FeatureID, c(" " = "_", "-" = "_"))) %>% 
    dplyr::select(SampleID, FeatureID, clr) %>%
    tidyr::spread(FeatureID, clr) %>% 
    dplyr::full_join(tse_meta(x), by = dplyr::join_by(SampleID == rownames))
}
```
```{r}
dfs <- taxa_subsets %>%
  purrr::map(melt_tse, .progress = TRUE)
```
# Taxa level analyses
generate a list of taxa, add the word GUT in front of the taxa
```{r}
taxa_rank_list <- function(x, y) {
  list_names <- colnames(x) %>%
    stringr::str_subset("GUT_") %>%
    rlang::set_names()
}
```
```{r}
taxa_rank_list <- dfs %>%
  purrr::imap(taxa_rank_list, .progress = TRUE)
```
## BMI
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_bmi_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_bmi_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_bmi_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for BMI")
  
```
## WaistCirc
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_waist_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_waist_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_waist_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for Waist Circ")
  
```
## Waist to hip ratio
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_whr_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_whr_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_whr_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for Waist Hip ratio")
```
Top 15 significant taxa
```{r}
whr<- df_lm_whr_results %>%
  dplyr::filter(qval_fdr < 0.05) %>%
  dplyr::arrange(qval_fdr) %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Genus:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("WHR", 15))

bmi <- df_lm_bmi_results %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Genus:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("BMI", 15))

waist <- df_lm_waist_results %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Genus:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("waist", 15))

hi <- rbind(whr, bmi, waist)

#define colors
my_colors <- c("waist" = "darkgreen", "BMI" = "orange", 
               "WHR" = "thistle")

q <- ggplot(data=hi, aes(x=taxa, y=estimate, fill = Origin)) +
  geom_bar(stat="identity", position = 'dodge') +
  scale_fill_manual(values=my_colors) +
  coord_flip() +
  theme_classic() +
  theme(axis.text.y = element_text(face = 'italic'))
q
```

#Common taxa
Agglomerate prevalent features
```{r}
#reads were filtered to detection threshold of 0.1% relative abundance, in 5% of samples
#This is to filter out rare taxa, to ensure we are anlaysing microbes that are likely to be real and not sequencing artefacts
subset_features_partial <-  purrr::partial(mia::agglomerateByPrevalence,
                                           x = tse,
                                           detection = args$detection,
                                           prevalence = args$prevalence,
                                           as_relative = TRUE) %>%
  purrr::possibly()
```
Subset to only Species
```{r}
taxa_subsets <- c("Family") %>%
  rlang::set_names() %>%
  purrr::map(subset_features_partial)
```
Melt tse to data frame
```{r}
melt_tse <- function(x, method = "clr") {
  mia::transformAssay(x, method = method, pseudocount = 1) %>% 
    mia::meltSE(add_row_data = TRUE, assay_name = method) %>%
    dplyr::mutate(FeatureID = glue::glue("GUT_{FeatureID}")) %>%
    dplyr::mutate(FeatureID = stringr::str_replace_all(FeatureID, c(" " = "_", "-" = "_"))) %>% 
    dplyr::select(SampleID, FeatureID, clr) %>%
    tidyr::spread(FeatureID, clr) %>% 
    dplyr::full_join(tse_meta(x), by = dplyr::join_by(SampleID == rownames))
}
```
```{r}
dfs <- taxa_subsets %>%
  purrr::map(melt_tse, .progress = TRUE)
```
# Taxa level analyses
generate a list of taxa, add the word GUT in front of the taxa
```{r}
taxa_rank_list <- function(x, y) {
  list_names <- colnames(x) %>%
    stringr::str_subset("GUT_") %>%
    rlang::set_names()
}
```
```{r}
taxa_rank_list <- dfs %>%
  purrr::imap(taxa_rank_list, .progress = TRUE)
```
## BMI
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_bmi_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_bmi_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_bmi_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for BMI")
  
```
## WaistCirc
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_waist_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_waist_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_waist_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for Waist Circ")
  
```
## Waist to hip ratio
From the taxa list, add rank, and generate linear model for each taxa, adjust for FDR using BH
```{r, message=FALSE}
df_lm_whr_results <- taxa_rank_list %>%
  purrr::map_df(~tibble::tibble(taxa = .x), .id = "rank") %>% 
  tidyr::gather(rank, taxa) %>%
  dplyr::mutate(results = purrr::map2(rank, taxa, ~lm_whr_model_for_taxon(df = dfs, term = .y, rank = .x))) %>%
  dplyr::mutate(results = purrr::map(results, ~broom::tidy(.x, conf.int = TRUE)))  %>%
  tidyr::unnest(results) %>%
  dplyr::filter(stringr::str_detect(term, "term")) %>%
  dplyr::mutate(qval_fdr = p.adjust(p.value, method = "BH"))
```
Results
```{r}
df_lm_whr_results %>%
  dplyr::arrange(p.value) %>% 
  dplyr::filter(qval_fdr < 0.05) %>%
  DT::datatable(caption = "Linear model for Waist Hip ratio")
```
Top 15 significant taxa
```{r}
whr<- df_lm_whr_results %>%
  dplyr::filter(qval_fdr < 0.05) %>%
  dplyr::arrange(qval_fdr) %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Family:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("WHR", 15))

bmi <- df_lm_bmi_results %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Family:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("BMI", 15))

waist <- df_lm_waist_results %>%
  dplyr::slice(1:15) %>%
  dplyr::select(taxa,estimate,conf.low, conf.high, p.value, qval_fdr) %>%
  dplyr::mutate(across(everything(),~ gsub("GUT_Family:","", .))) %>%
  dplyr::mutate_at(c(2:6), as.numeric) %>%
  dplyr::mutate_at(c(2:4),round,3) %>%
  dplyr::mutate_at(c(5:6), round, 3) %>%
  dplyr::mutate(Origin = rep("waist", 15))

hi <- rbind(whr, bmi, waist)

#define colors
my_colors <- c("waist" = "darkgreen", "BMI" = "orange", 
               "WHR" = "thistle")

q <- ggplot(data=hi, aes(x=taxa, y=estimate, fill = Origin)) +
  geom_bar(stat="identity", position = 'dodge') +
  scale_fill_manual(values=my_colors) +
  coord_flip() +
  theme_classic() +
  theme(axis.text.y = element_text(face = 'italic'))
q

```

Species
```{r}
tse_species <- mia::agglomerateByPrevalence(tse, rank = "Species",
                                              detection = 0.1/100,
                                              prevalence = 5/100,
                                              as_relative = TRUE)
```
ANCOM-BC2 for BMI
```{r}
pcol <- glue::glue("q_BMI")
```
```{r}
ancombc2 <- tse_species %>%
  ANCOMBC::ancombc2(data = .,
                    fix_formula = c("BL_AGE", "MEN", "PREVAL_DIAB", "CURR_SMOKE",
                                    "PREVAL_CVD", "ALKI2_FR02", "HFC",
                                    "Q57", "BMI") %>%
                      stringr::str_c(collapse = "+"),
                    assay.type = "counts",
                    p_adj_method = "fdr",
                    verbose = TRUE,
                    neg_lb = TRUE)
```
```{r}
BMI_df <- ancombc2$res %>%
  dplyr::select(taxon, contains("BMI")) %>%
  dplyr::arrange(!!rlang::sym(pcol))
```

```{r}
BMI_df %>%
  round_numeric_columns() %>% 
  dplyr::filter(q_BMI < 0.05) %>%
  DT::datatable(caption = "BMI")
```
ANCOM-BC2 for waist circ
```{r}
pcol <- glue::glue("q_VYOTARO")
```
```{r}
ancombc2 <- tse_species %>%
  ANCOMBC::ancombc2(data = .,
                    fix_formula = c("BL_AGE", "MEN", "PREVAL_DIAB", "CURR_SMOKE",
                                    "PREVAL_CVD", "ALKI2_FR02", "HFC",
                                    "Q57", "VYOTARO") %>%
                      stringr::str_c(collapse = "+"),
                    assay.type = "counts",
                    p_adj_method = "fdr",
                    verbose = TRUE,
                    neg_lb = TRUE)
```
```{r}
waist_df <- ancombc2$res %>%
  dplyr::select(taxon, contains("VYOTARO")) %>%
  dplyr::arrange(!!rlang::sym(pcol))
```

```{r}
waist_df %>%
  round_numeric_columns() %>% 
  dplyr::filter(q_VYOTARO < 0.05) %>%
  DT::datatable(caption = "VYOTARO")
```
ANCOM-BC2 for WHR
```{r}
pcol <- glue::glue("q_WHR")
```
```{r}
ancombc2 <- tse_species %>%
  ANCOMBC::ancombc2(data = .,
                    fix_formula = c("BL_AGE", "MEN", "PREVAL_DIAB", "CURR_SMOKE",
                                    "PREVAL_CVD", "ALKI2_FR02", "HFC",
                                    "Q57", "WHR") %>%
                      stringr::str_c(collapse = "+"),
                    assay.type = "counts",
                    p_adj_method = "fdr",
                    verbose = TRUE,
                    neg_lb = TRUE)
```
```{r}
WHR_df <- ancombc2$res %>%
  dplyr::select(taxon, contains("WHR")) %>%
  dplyr::arrange(!!rlang::sym(pcol))
```

```{r}
WHR_df %>%
  round_numeric_columns() %>% 
  dplyr::filter(q_WHR < 0.05) %>%
  DT::datatable(caption = "WHR")
```
```{r}
#ANCOMBC
sig.taxa <- BMI_df %>%
  dplyr::slice(1:10) %>%
  select(taxon)

waist <- waist_df %>% 
  dplyr::right_join(sig.taxa, join_by(taxon == taxon)) %>%
  rename_at('lfc_VYOTARO', ~'lfc') %>%
  dplyr::mutate(Origin = rep("waist", 10)) %>%
  dplyr::select(taxon, lfc, Origin)

bmi <- BMI_df %>%
  dplyr::right_join(sig.taxa, join_by(taxon == taxon)) %>%
  rename_at('lfc_BMI', ~'lfc') %>%
  dplyr::mutate(Origin = rep("BMI", 10)) %>%
  dplyr::select(taxon, lfc, Origin)
whr <- WHR_df %>%
  dplyr::right_join(sig.taxa, join_by(taxon == taxon)) %>%
  rename_at('lfc_WHR', ~'lfc') %>%
  dplyr::mutate(Origin = rep("WHR", 10))%>%
  dplyr::select(taxon, lfc, Origin)
hi <- rbind(whr, bmi, waist)

#define colors
my_colors <- c("waist" = "darkgreen", "BMI" = "orange", 
               "WHR" = "pink")

p <- ggplot(data=hi, aes(x=taxon, y=lfc, fill = Origin)) +
  geom_bar(stat="identity", position = 'dodge') +
  scale_fill_manual(values=my_colors) +
  coord_flip() +
  theme_classic() +
  theme(axis.text.y = element_text(face = 'italic'))
p

```

